### PR TITLE
feat(test runner): allow to pass arbitrary location to test.step

### LINKED ISSUE #30160: [Feature]: allow to pass arbitrary location to test.step
### 🚀 Feature Request

Currently `test.step` method binds location in file to the place of it's call. 
When `test.step` is wrapped with helper methods, it produces the same location for different steps.
Suggestion is to allow to pass arbitrary location to `test.step`. See example below.

### Example

Imagine I have a `helper.ts` where I define step helper for my tests:
```ts
export async function clickButton(text: string) {
  return test.step(`User clicks button "${text}"`, async () => {
    // ...
  }); 
}
```
Now I run the following test from `./e2e/test.ts`:
```ts
test('test', async ({}) => {
  await clickButton('foo');
  await clickButton('bar');
  await clickButton('baz');
});
```
In the report I get all steps with the same location:

It would be great to see more meaningful locations in the report, e.g.:
```
User clicks button "foo"— e2e/test.ts:2
User clicks button "bar"— e2e/test.ts:3
User clicks button "baz"— e2e/test.ts:4
```

If I were able to provide a custom location to `test.step`, I would achieve it. Example:
```ts
import { Location } from '@playwright/test/reporter';
import errorStackParser from 'error-stack-parser';

export async function clickButton(text: string) {
  const location = getStepLocation();
  return test.step(`User clicks button "${text}"`, async () => {
    // ...
  }, { location }); 
}

function getStepLocation(): Location {
  const frame = errorStackParser.parse(new Error()).find((frame) => frame.fileName.endsWith('.test.ts'));
  return {
      file: frame.getFileName(),
      line: frame.getLineNumber(),
      column: frame.getColumnNumber(),
  };
}
```

### Motivation

In general this feature provides more opportunities for writing high order step helpers.

Personally, I need it for managing `playwright-bdd` project. I'm wrapping `test.step` with `Given / When / Then` helpers to have correct step titles in the report:
```ts
const Given = (text: string) => test.step(`Given ${text}`, () => { ... });
const When = (text: string) => test.step(`When ${text}`, () => { ... });
const Then = (text: string) => test.step(`Then ${text}`, () => { ... });
```
Before Playwright 1.43 [I've used](https://github.com/vitalets/playwright-bdd/blob/v6.0.3/src/playwright/testTypeImpl.ts#L43) private `testInfo._runAsStep` method to run step with a custom location. Since 1.43 `testInfo._runAsStep` was replaced with a more complex logic and now I need [more tricky hacks](https://github.com/vitalets/playwright-bdd/blob/dd1f38d26b099ea5ff0bb38155947452c38f0a18/src/playwright/testTypeImpl.ts#L54) to achieve the same.

### PR DESCRIPTION
**Pull request by osohyun0224**
Fixes https://github.com/microsoft/playwright/issues/30160

### Description:
This pull request introduces the ability to specify custom locations for test steps in Playwright. By enabling the provision of arbitrary locations to the test.step method, it resolves the limitation where helper methods obfuscate the original call site, providing more accurate and meaningful location data in test reports.

### Motivation:
To enhance the utility and clarity of test reports in Playwright. Specifically, it addresses the need to trace test steps back to their precise location in the code, which is especially important when steps are abstracted in helper functions. This feature is crucial for maintaining accurate documentation and facilitating debugging processes.

### Changes:
Added functionality to pass a custom location object to test.step.

### Expected Outcome:
This PR is expected to significantly improve the precision and usefulness of diagnostic data in test reports by allowing specific locations within helper functions to be accurately documented. It facilitates better tracking of test executions and simplifies the debugging process, making it easier for developers to understand and address issues within complex tests.

### References:
Closes https://github.com/microsoft/playwright/issues/30160 - "[Feature]: allow to pass arbitrary location to test.step"

**Code Check**
I conducted tests on this new feature by integrating it into some existing test codes, and it worked well. I will attach the code used for testing and a screenshot showing the successful outcome.

<details>
<summary>toggle dropdown</summary>
<div markdown="1">

```
import type { Location } from '../../../packages/playwright/types/testReporter'
...
test('should respect the back button', async ({ page }) => {
    await page.locator('.todo-list li .toggle').nth(1).check();
    await checkNumberOfCompletedTodosInLocalStorage(page, 1);
...
    await test.step('Showing active items', async () => {
      await page.getByRole('link', { name: 'Active' }).click();
    }, {location});
```

image
</div>
</details>

### CODE REVIEW

File: examples/todomvc/tests/integration.spec.ts [hunk 1]
```diff
@@ -2,7 +2,6 @@
 
 import { test, expect } from '@playwright/test';
 import type { Page } from '@playwright/test';
-
```
Comment (dgozman) [inline examples/todomvc/tests/integration.spec.ts:5]: Why this stray change? Let's revert it.

File: packages/playwright/src/common/fixtures.ts [hunk 2]
```diff
@@ -23,7 +23,7 @@ import type { FixturesWithLocation } from './config';
 export type FixtureScope = 'test' | 'worker';
 type FixtureAuto = boolean | 'all-hooks-included';
 const kScopeOrder: FixtureScope[] = ['test', 'worker'];
-type FixtureOptions = { auto?: FixtureAuto, scope?: FixtureScope, option?: boolean, timeout?: number | undefined, title?: string, box?: boolean };
+type FixtureOptions = { auto?: FixtureAuto, scope?: FixtureScope, option?: boolean, timeout?: number | undefined, title?: string, box?: boolean, location?: Location };
```
Comment (dgozman) [inline packages/playwright/src/common/fixtures.ts:26]: It seems like any fixtures changes are unrelated to `test.step` locations. I think we should revert this one.

File: utils/generate_types/overrides-test.d.ts [hunk 3]
```diff
@@ -128,7 +128,7 @@ export interface TestType<TestArgs extends KeyValue, WorkerArgs extends KeyValue
   afterAll(inner: (args: TestArgs & WorkerArgs, testInfo: TestInfo) => Promise<any> | any): void;
   afterAll(title: string, inner: (args: TestArgs & WorkerArgs, testInfo: TestInfo) => Promise<any> | any): void;
   use(fixtures: Fixtures<{}, {}, TestArgs, WorkerArgs>): void;
-  step<T>(title: string, body: () => T | Promise<T>, options?: { box?: boolean }): Promise<T>;
+  step<T>(title: string, body: () => T | Promise<T>, options?: { box?: boolean, location?: Location }): Promise<T>;
```
Comment (dgozman) [inline utils/generate_types/overrides-test.d.ts:131]: In addition to this, you have to update documentation at `docs/src/test-api/class-test.md`.

**Review by dgozman — COMMENTED** [review]
Thank you for the PR! We will need a test for this - take a look at `test-step.spec.ts` for inspiration.

File: examples/todomvc/tests/integration.spec.ts (hunk 1 above)
Comment (osohyun0224) [reply examples/todomvc/tests/integration.spec.ts]: That's right. I think I committed by mistake. I'll correct it.

**Comment by osohyun0224** [comment]
@dgozman hello:)

First of all, thank you so much for actively reviewing my PR before commenting! I think it was an opportunity for me to improve further thanks to your review.

I accepted your review and reverted the entire implementation, adding test code to the `test-step.steps.ts` file and `class-test.md`. I wrote a description of the features in.

Can I ask you to review again focusing on this part? Thanks again for reviewing my PR.

**Review by dgozman — COMMENTED** [review]
The code looks good, but docs need some work, and tests do not pass. Thank you for working on this!

File: docs/src/test-api/class-test.md [hunk 4]
```diff
@@ -1710,6 +1710,12 @@ Step body.
 
 Whether to box the step in the report. Defaults to `false`. When the step is boxed, errors thrown from the step internals point to the step call site. See below for more details.
 
+### option: Test.step.location
+* since: v1.50
+- `location` <[Object]>
+
+Specifies a custom location for a test step as { file, line, column }, enabling precise error reporting within user code.
+
```
Comment (dgozman) [inline docs/src/test-api/class-test.md:1718]:
```suggestion
### option: Test.step.location
* since: v1.48
- `location` <[Location]>

Specifies a custom location for the step to be shown in test reports. By default, location of the [`method: Test.step`] call is shown.

```

**Review by dgozman — COMMENTED** [review]
Looks great, just a single suggestion for the docs.

File: docs/src/test-api/class-test.md (hunk 4 above)
Comment (osohyun0224) [reply docs/src/test-api/class-test.md]: @dgozman Thanks for the quick response to my modified commit! I will immediately reflect this part of your answer :)

**Review by dgozman — APPROVED** [review]
Thank you for the PR, looks great!

File: packages/playwright/types/test.d.ts [hunk 5]
```diff
@@ -4703,7 +4703,7 @@ export interface TestType<TestArgs extends KeyValue, WorkerArgs extends KeyValue
    * @param body Step body.
    * @param options
    */
-  step<T>(title: string, body: () => T | Promise<T>, options?: { box?: boolean }): Promise<T>;
+  step<T>(title: string, body: () => T | Promise<T>, options?: { box?: boolean, location?: Location }): Promise<T>;
```
Comment (jan-molak) [inline packages/playwright/types/test.d.ts:4706]: Please note that `Location` here corresponds to the global `lib-dom` `Location`, as opposed to the Playwright's [`Location`](https://github.com/microsoft/playwright/blob/9fc195bff5cd4a5dcb4337d8c9a4400683c9c62b/packages/playwright/types/testReporter.d.ts#L325-L340) interface.

I'd suggest moving the definition of Playwright's `Location` from `testReporter.d.ts` to `playwright/types/test.d.ts` since it's now part of the public API.

@dgozman - what's your view?

Comment (osohyun0224) [reply packages/playwright/types/test.d.ts:4706]: Looking at it along with the issues you mentioned, I think it's right to change like you said.
The maintainers fixed the bug according to your issue. Thank you for the review. 👍🏻👍🏻
